File: 백준/Gold/3190. 뱀/뱀.py
n = int(input())
k = int(input()) #사과개수
arr = [[0]*(n+1) for _ in range(n+1)]
for _ in range(k):
    x,y = map(int, input().split())
    arr[x][y] = 1
L = int(input())
info = []
for _ in range(L):
    X,C = map(str, input().split())
    info.append((int(X),C))

#동 남 서 북 (오른쪽 머리)
dx = [0,1,0,-1]
dy = [1,0,-1,0]

def turn(dir, c):
    if c == 'L':
        dir = (dir -1)%4
    else:
        dir = (dir+1)%4
    return dir


def simulate():
    x,y = 1,1
    dir = 0
    arr[x][y] = 2 #뱀의 머리가 있는곳
    time = 0
    idx = 0 #회전정보
    q = [(x,y)] #뱀의 몸통이 차지하고 있는 위치 정보(꼬리가 앞)
    while True:
        nx = x+ dx[dir]
        ny = y + dy[dir]
        # 맵안, 뱀위 몸통이 없는곳
        if 1<=nx and nx <=n and 1<=ny and ny<=n and arr[nx][ny]!=2:
            if arr[nx][ny] == 0:
                arr[nx][ny] = 2
                q.append((nx,ny))
                px,py = q.pop(0)
                arr[px][py] = 0
            if arr[nx][ny] == 1:
                arr[nx][ny] = 2
                q.append((nx,ny))
        else:
            time +=1
            break
        x,y = nx,ny
        time +=1
        if idx < L and time == info[idx][0]:
            dir = turn(dir, info[idx][1])
            idx += 1
    return time


print(simulate())

File: 백준/Gold/3190. 뱀/뱀.java
import java.util.*;
import java.io.*;

class Main {
    static int N;
    static boolean[][] apples;
    static Map<Integer, Character> directionChanges = new HashMap<>();
    static int[]dx = {0, 1, 0, -1};
    static int[] dy = {1,0,-1,0};
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        // 보드 크기
        N = Integer.parseInt(br.readLine());

        // 사과 개수
        int K = Integer.parseInt(br.readLine());

        // 사과 위치 저장 (1-based)
        apples = new boolean[N + 1][N + 1];
        for (int i = 0; i < K; i++) {
            st = new StringTokenizer(br.readLine());
            int row = Integer.parseInt(st.nextToken());
            int col = Integer.parseInt(st.nextToken());
            apples[row][col] = true;
        }

        // 방향 변환 횟수
        int L = Integer.parseInt(br.readLine());

        // 방향 전환 정보 저장
        for (int i = 0; i < L; i++) {
            st = new StringTokenizer(br.readLine());
            int time = Integer.parseInt(st.nextToken());
            char dir = st.nextToken().charAt(0);
            directionChanges.put(time, dir);
        }

        int res = simulate();
        System.out.println(res);

    }
    public static int simulate(){
        Deque<int[]> snake = new LinkedList<>();
        snake.add(new int[]{1,1});

        boolean[][] visited = new boolean[N+1][N+1];
        visited[1][1]= true;

        int time = 0;
        int dir = 0;
        int x = 1,y=1;
        while (true){
            time++;
            int nx = x+dx[dir];
            int ny = y + dy[dir];
            if(nx<1 || nx > N || ny < 1||ny>N ||visited[nx][ny]) break;

            snake.addFirst(new int[]{nx,ny});
            visited[nx][ny] = true;

            if(apples[nx][ny]){
                apples[nx][ny] = false;
            }else{
                int[] tail = snake.removeLast();
                visited[tail[0]][tail[1]]= false;
            }

            if(directionChanges.containsKey(time)){
                char turn = directionChanges.get(time);
                if(turn == 'L'){
                    dir = (dir+3)%4;
                }else if(turn == 'D'){
                    dir = (dir+1)%4;
                }
            }
            x = nx;
            y = ny;
        }
        return time;
    }
}